testsuite/forge/src/backend/k8s/kube_api.rs

```diff
@@ -61,299 +61,148 @@ where
 
 #[cfg(test)]
 pub mod mocks {
-    use super::*;
+    use super::ReadWrite;
     use crate::Result;
+    use aptos_infallible::Mutex;
     use async_trait::async_trait;
     use hyper::StatusCode;
-    use k8s_openapi::api::{
-        apps::v1::StatefulSet,
-        core::v1::{ConfigMap, Namespace, PersistentVolumeClaim, Pod, Secret, Service},
+    use k8s_openapi::Metadata;
+    use kube::{
+        api::{ObjectMeta, PostParams},
+        error::ErrorResponse,
+        Error as KubeError,
     };
-    use kube::{api::PostParams, error::ErrorResponse, Error as KubeError};
+    use std::{collections::BTreeMap, sync::Arc};
 
-    // Mock StatefulSet API
-
-    pub struct MockStatefulSetApi {
-        stateful_set: StatefulSet,
-    }
-
-    impl MockStatefulSetApi {
-        pub fn from_stateful_set(stateful_set: StatefulSet) -> Self {
-            MockStatefulSetApi { stateful_set }
-        }
-    }
-
-    #[async_trait]
-    impl ReadWrite<StatefulSet> for MockStatefulSetApi {
-        async fn get(&self, name: &str) -> Result<StatefulSet, KubeError> {
-            if self.stateful_set.metadata.name == Some(name.to_string()) {
-                return Ok(self.stateful_set.clone());
-            }
-            return Err(KubeError::Api(ErrorResponse {
-                status: "failed".to_string(),
-                message: format!(
-                    "StatefulSet with name {} could not be found in {:?}",
-                    name, self.stateful_set
-                ),
-                reason: "not_found".to_string(),
-                code: 404,
-            }));
-        }
-
-        async fn create(
-            &self,
-            _pp: &PostParams,
-            stateful_set: &StatefulSet,
-        ) -> Result<StatefulSet, KubeError> {
-            if self.stateful_set.metadata.name == stateful_set.metadata.name {
-                return Err(KubeError::Api(ErrorResponse {
-                    status: "failed".to_string(),
-                    message: format!(
-                        "StatefulSet with same name already exists in {:?}",
-                        self.stateful_set
-                    ),
-                    reason: "already_exists".to_string(),
-                    code: 409,
-                }));
-            }
-            Ok(self.stateful_set.clone())
-        }
+    /// Generic k8s resource mock API where resource names are unique. Use it to mock namespaced resources or cluster-wide resources, but
+    /// not resources across multiple namespaces.
+    #[derive(Clone)]
+    pub struct MockK8sResourceApi<T> {
```

Comment:
And you can get rid of the template here too!

Although this is also anyways much nicer than what we had before 🙏